File: sdks/python/evalsmith/trace.py
"""Trace and Span context managers for manual instrumentation."""

from __future__ import annotations

import time
import uuid
from datetime import datetime, timezone
from typing import Any, Callable


def _iso_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _ts_to_iso(ts: float) -> str:
    if ts == 0:
        return ""
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


class Span:
    """Represents a single execution step within a Trace."""

    def __init__(
        self,
        name: str,
        span_type: str = "custom",
        parent_span_id: str | None = None,
        on_exit: Callable[[Span], None] | None = None,
    ):
        self.span_id = f"sp_{uuid.uuid4().hex[:12]}"
        self.parent_span_id = parent_span_id
        self.name = name
        self.span_type = span_type
        self.status = "ok"
        self.start_time: float = 0
        self.end_time: float = 0
        self.input: Any = None
        self.output: Any = None
        self.model: str | None = None
        self.token_input: int = 0
        self.token_output: int = 0
        self.cost_usd: float = 0.0
        self.metadata: dict = {}
        self.events: list[dict] = []
        self.error_message: str | None = None
        self.children: list[Span] = []
        self._on_exit = on_exit

    def set_input(self, value: Any):
        self.input = value
        return self

    def set_output(self, value: Any):
        self.output = value
        return self

    def set_metadata(self, **values: Any):
        self.metadata.update(values)
        return self

    def set_model(self, model: str, token_input: int = 0, token_output: int = 0, cost_usd: float = 0.0):
        self.model = model
        self.token_input = token_input
        self.token_output = token_output
        self.cost_usd = cost_usd
        return self

    def set_error(self, message: str):
        self.status = "error"
        self.error_message = message
        return self

    def add_event(self, name: str, attributes: dict | None = None):
        self.events.append(
            {
                "name": name,
                "timestamp": _iso_now(),
                "attributes": attributes or {},
            }
        )
        return self

    def to_dict(self) -> dict:
        payload: dict[str, Any] = {
            "span_id": self.span_id,
            "parent_span_id": self.parent_span_id,
            "name": self.name,
            "span_type": self.span_type,
            "status": self.status,
            "start_time": _ts_to_iso(self.start_time),
            "end_time": _ts_to_iso(self.end_time),
            "input": self.input,
            "output": self.output,
            "metadata": self.metadata,
            "events": self.events,
            "metrics": {
                "model": self.model,
                "token_input": self.token_input,
                "token_output": self.token_output,
                "cost_usd": self.cost_usd,
            },
        }
        if self.error_message:
            payload["error_message"] = self.error_message
        return payload

    def __enter__(self):
        self.start_time = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.end_time = time.time()
        if exc_type is not None:
            self.set_error(str(exc_val))
        if self._on_exit is not None:
            self._on_exit(self)
        return False


class Trace:
    """Represents a complete agent execution trace."""

    def __init__(self, name: str, tags: list[str] | None = None, metadata: dict | None = None):
        self.trace_id = f"tr_{uuid.uuid4().hex[:12]}"
        self.name = name
        self.tags = tags or []
        self.metadata = metadata or {}
        self.spans: list[Span] = []
        self._span_stack: list[Span] = []
        self.start_time: float = 0
        self.end_time: float = 0

    def span(self, name: str, span_type: str = "custom") -> Span:
        """Create a child span. Use as a context manager."""

        parent = self._span_stack[-1] if self._span_stack else None
        span = Span(
            name=name,
            span_type=span_type,
            parent_span_id=parent.span_id if parent else None,
            on_exit=self._on_span_exit,
        )
        if parent is not None:
            parent.children.append(span)
        self.spans.append(span)
        self._span_stack.append(span)
        return span

    def to_dict(self) -> dict:
        return {
            "trace_id": self.trace_id,
            "name": self.name,
            "tags": self.tags,
            "metadata": self.metadata,
            "spans": [span.to_dict() for span in self.spans],
        }

    def __enter__(self):
        self.start_time = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.end_time = time.time()
        if exc_type is not None and self._span_stack:
            self._span_stack[-1].set_error(str(exc_val))

        import evalsmith

        client = evalsmith.get_client()
        client.enqueue_trace(self.to_dict())
        return False

    def _on_span_exit(self, span: Span) -> None:
        if self._span_stack and self._span_stack[-1].span_id == span.span_id:
            self._span_stack.pop()
            return
        self._span_stack = [item for item in self._span_stack if item.span_id != span.span_id]

File: sdks/go/evalsmith/trace.go
package evalsmith

import (
	"context"
	"fmt"
	"time"
)

type BatchIngestRequest struct {
	Traces []TracePayload `json:"traces"`
}

type TracePayload struct {
	TraceID  string         `json:"trace_id,omitempty"`
	Name     string         `json:"name"`
	Tags     []string       `json:"tags,omitempty"`
	Metadata map[string]any `json:"metadata,omitempty"`
	Spans    []SpanPayload  `json:"spans"`
}

type SpanPayload struct {
	SpanID       string           `json:"span_id,omitempty"`
	ParentSpanID *string          `json:"parent_span_id,omitempty"`
	Name         string           `json:"name"`
	SpanType     string           `json:"span_type"`
	Status       string           `json:"status"`
	StartTime    string           `json:"start_time"`
	EndTime      string           `json:"end_time"`
	Input        any              `json:"input,omitempty"`
	Output       any              `json:"output,omitempty"`
	Metrics      map[string]any   `json:"metrics,omitempty"`
	Metadata     map[string]any   `json:"metadata,omitempty"`
	Events       []map[string]any `json:"events,omitempty"`
}

func NewTrace(name string, tags []string, metadata map[string]any) *Trace {
	return &Trace{
		TraceID:  "tr_" + time.Now().UTC().Format("20060102150405.000000"),
		Name:     name,
		Tags:     tags,
		Metadata: metadata,
		Spans:    []SpanPayload{},
	}
}

// Trace represents a full execution trace with nested spans.
type Trace struct {
	TraceID  string
	Name     string
	Tags     []string
	Metadata map[string]any
	Spans    []SpanPayload
}

// AddSpan appends a flat span (no parent tracking).
func (t *Trace) AddSpan(name string, spanType string, input any, output any, metadata map[string]any) {
	started := time.Now().UTC()
	ended := time.Now().UTC()
	t.Spans = append(t.Spans, SpanPayload{
		SpanID:    "sp_" + started.Format("150405.000000"),
		Name:      name,
		SpanType:  spanType,
		Status:    "ok",
		StartTime: started.Format(time.RFC3339Nano),
		EndTime:   ended.Format(time.RFC3339Nano),
		Input:     input,
		Output:    output,
		Metadata:  metadata,
		Metrics:   map[string]any{},
	})
}

// StartSpan creates a nested span with proper parent-child linkage.
func (t *Trace) StartSpan(name string, spanType string, parentSpanID string) *Span {
	spanID := fmt.Sprintf("sp_%d", time.Now().UnixNano())
	var parent *string
	if parentSpanID != "" {
		parent = &parentSpanID
	}
	return &Span{
		trace:    t,
		SpanID:   spanID,
		ParentID: parent,
		Name:     name,
		SpanType: spanType,
		Status:   "ok",
		Started:  time.Now().UTC(),
		Metrics:  map[string]any{},
		Metadata: map[string]any{},
	}
}

// Span is an in-progress span builder. Call End() to finalize and add to Trace.
type Span struct {
	trace    *Trace
	SpanID   string
	ParentID *string
	Name     string
	SpanType string
	Status   string
	Started  time.Time
	Input    any
	Output   any
	Metrics  map[string]any
	Metadata map[string]any
	Error    string
}

func (s *Span) SetInput(input any)            { s.Input = input }
func (s *Span) SetOutput(output any)          { s.Output = output }
func (s *Span) SetMetric(key string, val any) { s.Metrics[key] = val }
func (s *Span) SetError(msg string)           { s.Status = "error"; s.Error = msg }

// End finalizes the span and appends it to the parent trace.
func (s *Span) End() {
	ended := time.Now().UTC()
	s.trace.Spans = append(s.trace.Spans, SpanPayload{
		SpanID:       s.SpanID,
		ParentSpanID: s.ParentID,
		Name:         s.Name,
		SpanType:     s.SpanType,
		Status:       s.Status,
		StartTime:    s.Started.Format(time.RFC3339Nano),
		EndTime:      ended.Format(time.RFC3339Nano),
		Input:        s.Input,
		Output:       s.Output,
		Metrics:      s.Metrics,
		Metadata:     s.Metadata,
	})
}

// StartChild creates a child span of this span.
func (s *Span) StartChild(name string, spanType string) *Span {
	return s.trace.StartSpan(name, spanType, s.SpanID)
}

func (t *Trace) Payload() TracePayload {
	return TracePayload{
		TraceID:  t.TraceID,
		Name:     t.Name,
		Tags:     t.Tags,
		Metadata: t.Metadata,
		Spans:    t.Spans,
	}
}

func (c *Client) IngestTrace(ctx context.Context, trace *Trace) error {
	return c.IngestTraces(ctx, []TracePayload{trace.Payload()})
}

func (c *Client) IngestTraces(ctx context.Context, traces []TracePayload) error {
	return c.PostJSON(ctx, c.traceURL, "/api/v1/traces", BatchIngestRequest{Traces: traces}, nil)
}
